cmake/utils.cmake

```diff
@@ -0,0 +1,58 @@
+cmake_minimum_required(VERSION 3.2)
+
+function(FindAndSelectClangCompiler)
+  if(DEFINED ENV{LLVM_INSTALL_PREFIX})
+    set(LLVM_INSTALL_PREFIX $ENV{LLVM_INSTALL_PREFIX} PARENT_SCOPE)
+  endif()
+
+  if(DEFINED LLVM_INSTALL_PREFIX)
+    list(APPEND FINDPACKAGE_LLVM_HINTS "${LLVM_INSTALL_PREFIX}/lib/cmake/llvm/")
+    list(APPEND FINDPACKAGE_LLVM_HINTS "${LLVM_INSTALL_PREFIX}/share/llvm/cmake/")
+
+    message(STATUS "Using LLVM_INSTALL_PREFIX hints for find_package(LLVM): ${FINDPACKAGE_LLVM_HINTS}")
+  endif()
+
+  if(DEFINED WIN32)
+    set(executable_extension ".exe")
+  else()
+    set(executable_extension "")
+  endif()
+
+  # it is important to avoid re-defining these variables if they have been already
+  # set or you risk ending up in a configure loop!
+  if(NOT DEFINED CMAKE_C_COMPILER)
+    if(DEFINED LLVM_INSTALL_PREFIX)
+      set(CMAKE_C_COMPILER "${LLVM_INSTALL_PREFIX}/bin/clang${executable_extension}"
+        CACHE PATH "Path to clang binary." PARENT_SCOPE)
+    else()
+      set(CMAKE_C_COMPILER "clang" PARENT_SCOPE)
+    endif()
+  endif()
+
+  if(NOT DEFINED CMAKE_CXX_COMPILER)
+    if(DEFINED LLVM_INSTALL_PREFIX)
+      set(CMAKE_CXX_COMPILER "${LLVM_INSTALL_PREFIX}/bin/clang++${executable_extension}"
+        CACHE PATH "Path to clang++ binary." PARENT_SCOPE)
+    else()
+      set(CMAKE_CXX_COMPILER "clang++${executable_extension}" PARENT_SCOPE)
+    endif()
+  endif()
+
+  if(NOT DEFINED CMAKE_ASM_COMPILER)
+    if(DEFINED LLVM_INSTALL_PREFIX)
+      set(CMAKE_ASM_COMPILER "${LLVM_INSTALL_PREFIX}/bin/clang++${executable_extension}"
+        CACHE PATH "Path to assembler (aka clang) binary." PARENT_SCOPE)
```

Comment:
With both `PARENT_SCOPE` and `CACHE`, this generates a mangled value on my system:
`"/home/dureuill/build/llvm/llvm/bin/clang++;CACHE;PATH;"Path to assembler (aka clang) binary."`

I believe this is because in CMake, a `set` value is either a local variable (and you can use `PARENT_SCOPE`), or a cache variable (and you cannot use `PARENT_SCOPE`) (see [cmake documentation](https://cmake.org/cmake/help/v3.9/command/set.html))

Since `CMAKE_ASM_COMPILER` is a cache variable, you should keep the `CACHE`, and leave out `PARENT_SCOPE`.
This fixes the problem on my machine.

This also applies to other CACHE variables in this file (CMAKE_CXX_COMPILER and CMAKE_C_COMPILER).

Reply:
I will try this change.